File: Difficulty: Easy/Minimum Cost of ropes/minimum-cost-of-ropes.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    // Function to return the minimum cost of connecting the ropes.
    long long minCost(vector<long long>& arr) {
        if(arr.size() <= 1)
            return 0;
            
        priority_queue<long long, vector<long long>, greater<long long>> pq(arr.begin(), arr.end());
        long long cost = 0;
        while(!pq.empty()) {
            int num1 = pq.top(); pq.pop();
            int num2 = pq.top(); pq.pop();
            cost += (num1 + num2);
            if(!pq.empty())
                pq.push(num1 + num2);
        }
        return cost;
    }
};


//{ Driver Code Starts.

int main() {
    long long t;
    cin >> t;
    cin.ignore();
    while (t--) {
        string input;
        long long num;
        vector<long long> a;

        getline(cin, input);
        stringstream s2(input);
        while (s2 >> num) {
            a.push_back(num);
        }
        Solution ob;
        cout << ob.minCost(a) << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Minimum Cost of ropes/minimum-cost-of-ropes.java
class Solution {
    public static int minCost(int[] arr) {
        if(arr.length == 1) return 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for(int num: arr) pq.add(num);
        int cost = 0;
        while(pq.size() > 1) {
            int r1 = pq.poll();
            int r2 = pq.poll();
            int sum = r1 + r2;
            cost += sum;
            pq.add(sum);
        }
        return cost;
    }
}